File: snies-backend/software_activities/domain/entities/software_activity.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from decimal import Decimal


@dataclass(frozen=True)
class SoftwareActivity:
    id: int | None
    year: int
    semester: int
    start_date: date | None
    end_date: date | None
    execution_place: str
    campus: str
    activity_name: str
    agreement_entity: str | None
    description: str | None
    cine_isced_name: str | None
    cine_field_detailed_id: str | None
    num_hours: int | None
    activity_type: str | None
    course_value: Decimal | None
    teacher_document_type: str | None
    teacher_document_number: str | None
    total_beneficiaries: int | None
    professors_count: int | None
    administrative_count: int | None
    external_people_count: int | None
    speaker_full_name: str | None
    speaker_origin: str | None
    speaker_company: str | None
    consultancy_entity_name: str | None
    consultancy_sector_id: str | None
    consultancy_value: Decimal | None
    evidence_event_planning: bool
    evidence_attendance_control: bool
    evidence_program_design_guide: bool
    evidence_audiovisual_record: bool


@dataclass(frozen=True)
class BeneficiaryBreakdown:
    id: int | None
    activity_id: int | None
    population: str  # students|graduates
    campus: str  # CÚCUTA|OCAÑA
    program: str
    level: str  # técnico|tecnólogo|profesional
    count: int

File: snies-frontend/modules/software_activities/types/software-activity.ts
export type BeneficiaryBreakdown = {
  population: string;
  campus: string;
  program: string;
  level: string;
  count: number;
};

export type SoftwareActivity = {
  id: number;
  career: string | null;
  year: number;
  semester: number;
  start_date: string | null;
  end_date: string | null;
  execution_place: string;
  campus: string;
  activity_name: string;
  agreement_entity: string | null;
  description: string | null;
  cine_isced_name: string | null;
  cine_field_detailed_id: string | null;
  num_hours: number | null;
  activity_type: string | null;
  course_value: string | null;
  teacher_document_type: string | null;
  teacher_document_number: string | null;
  total_beneficiaries: number | null;
  professors_count: number | null;
  administrative_count: number | null;
  external_people_count: number | null;
  speaker_full_name: string | null;
  speaker_origin: string | null;
  speaker_company: string | null;
  consultancy_entity_name: string | null;
  consultancy_sector_id: string | null;
  consultancy_value: string | null;
  evidence_event_planning: boolean;
  evidence_event_planning_file?: string | null;
  evidence_attendance_control: boolean;
  evidence_attendance_control_file?: string | null;
  evidence_program_design_guide: boolean;
  evidence_program_design_guide_file?: string | null;
  evidence_audiovisual_record: boolean;
  evidence_audiovisual_record_file?: string | null;
  created_at: string;
  beneficiary_breakdowns?: BeneficiaryBreakdown[];
};

export type CreateSoftwareActivityInput = {
  career?: string | null;
  year: number;
  semester: number;
  start_date?: string | null;
  end_date?: string | null;
  execution_place: string;
  campus: string;
  activity_name: string;
  agreement_entity?: string | null;
  description?: string | null;
  cine_isced_name?: string | null;
  cine_field_detailed_id?: string | null;
  num_hours?: number | null;
  activity_type?: string | null;
  course_value?: number | null;
  teacher_document_type?: string | null;
  teacher_document_number?: string | null;
  total_beneficiaries?: number | null;
  professors_count?: number | null;
  administrative_count?: number | null;
  external_people_count?: number | null;
  speaker_full_name?: string | null;
  speaker_origin?: string | null;
  speaker_company?: string | null;
  consultancy_entity_name?: string | null;
  consultancy_sector_id?: string | null;
  consultancy_value?: number | null;
  evidence_event_planning?: boolean;
  evidence_attendance_control?: boolean;
  evidence_program_design_guide?: boolean;
  evidence_audiovisual_record?: boolean;
  beneficiary_breakdowns?: BeneficiaryBreakdown[];
};

export type ImportSoftwareActivitiesResult = {
  created: number;
  skipped_empty_rows: number;
};

